### PR TITLE
feat(main): allow retries to be limited

### PR DESCRIPTION
**Pull request by oliversalzburg**

### CODE REVIEW

**Comment by oliversalzburg** [comment]
Working on fixing the coverage. I already noticed problems as well. Will post an update soon.

**Comment by oliversalzburg** [comment]
@voxpelli I think I need your help. husky prevents me from pushing, because jshint doesn't like the `.jshintrc`:

```
Running "lintspaces:files" (lintspaces) task
>> 3 lint free.

Running "jshint:files" (jshint) task
ERROR: Can't parse config file: .jshintrc
Error:SyntaxError: Unexpected token o
```

**Comment by oliversalzburg** [comment]
Ah, got that solved. Those are supposed to be symlinks. Now it fails at a different spot:

```
Running "dependency-check:files" (dependency-check) task
Fatal error: Object.keys called on non-object
```

**Comment by oliversalzburg** [comment]
I pushed with `--no-verfiy` tests are passing now. I would appreciate if you could have a look at everything. I'm not sure I fully understood the tests.

Let me know if I need to fix anything :)

**Comment by voxpelli** [comment]
@oliversalzburg Sorry for the `dependency-check` failure – I'm co-maintaining that module so would like to try and find out what the error was. Which Node are you running? Some `0.12` version r an io.js version?

File: index.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ var Retry = function (options) {
     retryMin: 0,
     retryBase: 1.2,
     retryExponent: 33,
+    retryLimit: -1,
```
Comment (voxpelli) [inline index.js]: Would prefer if the default was either `false` or `undefined` – more explicit about the intention of the value

File: index.js [hunk 2]
```diff
@@ -76,7 +77,13 @@ Retry.prototype._try = function () {
     self.retrying = undefined;
     self.abort = undefined;
 
-    return self._try();
+
+    if (self.options.retryLimit > -1 && self.failures >= self.options.retryLimit) {
+      self.end();
+      return Promise.reject(new Error('Retry limit reached'));
+    } else {
```
Comment (voxpelli) [inline index.js]: No need to have an `} else {` here – if the first `if` has returned than anything following it won't run anyway so just put the last return at the end of the method.

**Comment by oliversalzburg** [comment]
@voxpelli I'm on 0.12.7 on Windows 10

**Comment by oliversalzburg** [comment]
@voxpelli Thanks for the review. New version is up :)

**Comment by voxpelli** [comment]
Looks like the `dependency-check` issue has been reported: https://github.com/maxogden/dependency-check/issues/30

File: index.js [hunk 3]
```diff
@@ -76,6 +77,12 @@ Retry.prototype._try = function () {
     self.retrying = undefined;
     self.abort = undefined;
 
+
+    if (self.options.retryLimit > -1 && self.failures >= self.options.retryLimit) {
```
Comment (voxpelli) [inline index.js]: This check also needs to be updated now that the default has changed

Comment (oliversalzburg) [inline index.js]: Jeeze, it's getting late :P Sorry for missing that.

**Comment by oliversalzburg** [comment]
Very cool :) I'll start working on https://github.com/voxpelli/node-pg-pubsub/pull/6 again tomorrow at work. Thanks for the help!

**Comment by voxpelli** [comment]
Published as `0.2.3` – extended the tests a bit and as the result of that changed the code a bit as well. Also removed the `.end()` – not needed and slipped through my first review :)

**Comment by oliversalzburg** [comment]
Interesting. I didn't understand the difference between end()ing and plain stopping :\

**Comment by voxpelli** [comment]
Ending is for graceful and immediate shutdowns of applications, to ensure that no matter what happens nothing, not even a reset, will try to establish a new connection :)

**Pull request merged**